File: JavaScript/2029E.js
"use strict";

const fs = require('fs');

/**
 * Problem: Codeforces 2029E - Common Generator (JavaScript Version)
 * Logic:
 * 1. Precompute SPF up to 400,005 using a sieve.
 * 2. Identify distinct primes in the input array.
 * 3. If > 1 distinct primes exist, no solution (-1).
 * 4. If 0 or 1 distinct primes exist, use the identified prime (or 2) as candidate x.
 * 5. Verify reachability for all elements using the rule:
 *    - Even v: v >= 2x
 *    - Odd composite v: v - spf[v] >= 2x
 */

function main() {
    const buffer = fs.readFileSync(0);
    let offset = 0;

    function nextInt() {
        while (offset < buffer.length && (buffer[offset] < 48 || buffer[offset] > 57)) offset++;
        if (offset >= buffer.length) return null;
        let res = 0;
        while (offset < buffer.length && buffer[offset] >= 48 && buffer[offset] <= 57) {
            res = res * 10 + (buffer[offset] - 48);
            offset++;
        }
        return res;
    }

    const MAXA = 400005;
    const spf = new Int32Array(MAXA);
    for (let i = 2; i < MAXA; i++) spf[i] = i;
    for (let i = 2; i * i < MAXA; i++) {
        if (spf[i] === i) {
            for (let j = i * i; j < MAXA; j += i) {
                if (spf[j] === j) spf[j] = i;
            }
        }
    }

    const t = nextInt();
    if (t === null) return;

    let output = "";
    for (let tc = 0; tc < t; tc++) {
        const n = nextInt();
        const a = new Int32Array(n);
        let primeCandidate = -1;
        let multiPrimes = false;

        for (let i = 0; i < n; i++) {
            const v = nextInt();
            a[i] = v;
            if (spf[v] === v) {
                if (primeCandidate === -1) {
                    primeCandidate = v;
                } else if (primeCandidate !== v) {
                    multiPrimes = true;
                }
            }
        }

        if (multiPrimes) {
            output += "-1\n";
        } else {
            const x = (primeCandidate === -1 ? 2 : primeCandidate);
            let ok = true;
            for (let i = 0; i < n; i++) {
                const v = a[i];
                if (v === x) continue;
                if (v % 2 === 0) {
                    if (v < 2 * x) { ok = false; break; }
                } else {
                    // v is odd. Since multiPrimes check passed, v must be composite.
                    if (v - spf[v] < 2 * x) { ok = false; break; }
                }
            }
            output += (ok ? x : -1) + "\n";
        }

        // Flush output periodically to manage memory
        if (output.length > 16384) {
            process.stdout.write(output);
            output = "";
        }
    }
    process.stdout.write(output);
}

main();

File: CPP/2029E.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

/**
 * Problem: Codeforces 2029E - Common Generator
 * Logic:
 * 1. A generator x can only reach a prime y if x = y.
 * 2. If the array has multiple distinct primes, no solution exists.
 * 3. If there's one prime p, it must be the generator.
 * 4. If there are no primes, x = 2 is a valid generator for all composite numbers.
 * 5. A prime p generates v if:
 *    - v == p
 *    - v is even and v >= 2p
 *    - v is odd composite and v - spf[v] >= 2p
 */

const int MAXA = 400005;
int spf[MAXA];

void precompute() {
    for (int i = 2; i < MAXA; ++i) spf[i] = i;
    for (int i = 2; i * i < MAXA; ++i) {
        if (spf[i] == i) {
            for (int j = i * i; j < MAXA; j += i) {
                if (spf[j] == j) spf[j] = i;
            }
        }
    }
}

void solve() {
    int n;
    if (!(cin >> n)) return;
    vector<int> a(n);
    vector<int> primes;
    for (int i = 0; i < n; ++i) {
        cin >> a[i];
        if (spf[a[i]] == a[i]) {
            primes.push_back(a[i]);
        }
    }
    
    sort(primes.begin(), primes.end());
    primes.erase(unique(primes.begin(), primes.end()), primes.end());
    
    if (primes.size() > 1) {
        cout << -1 << "\n";
        return;
    }
    
    int x = (primes.empty() ? 2 : primes[0]);
    
    for (int v : a) {
        if (v == x) continue;
        
        // Check if x generates v
        if (v % 2 == 0) {
            // Even v can be reached if v >= 2x
            if (v < 2 * x) {
                cout << -1 << "\n";
                return;
            }
        } else {
            // Odd composite v can be reached if we can reach an even number 
            // that can jump to v. The jump must be via an odd divisor of v.
            // The jump from even z to odd v is most flexible if z = v - spf[v].
            // This z is even and must be >= 2x.
            if (v - spf[v] < 2 * x) {
                cout << -1 << "\n";
                return;
            }
        }
    }
    
    cout << x << "\n";
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    precompute();
    int t;
    if (!(cin >> t)) return 0;
    while (t--) {
        solve();
    }
    return 0;
}
